rusoto/core/src/signature.rs

```diff
@@ -229,6 +229,102 @@ impl SignedRequest {
         self.params = params;
     }
 
+
+    /// http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+    pub fn generate_presigned_url(&mut self, creds: &AwsCredentials) -> String {
+        debug!("Presigning request URL");
+
+        self.sign(creds);
+        let hostname = match self.hostname {
+            Some(ref h) => h.to_string(),
+            None => build_hostname(&self.service, &self.region),
+        };
+
+        let current_time = now_utc();
+        let current_time_fmted = current_time.strftime("%Y%m%dT%H%M%SZ").unwrap();
+        let current_time_fmted = format!("{}", &current_time_fmted);
+        let current_date = current_time.strftime("%Y%m%d").unwrap();
+
+        self.remove_header("x-amz-content-sha256");
+        self.remove_header("X-Amz-Content-Sha256");
+
+        self.remove_header("X-Amz-Date");
+        self.remove_header("x-amz-date");
+
+        self.remove_header("content-type");
+        self.remove_header("Content-Type");
+
+        self.remove_header("X-Amz-Algorithm");
+        self.remove_header("x-amz-algorithm");
+        self.params.put("X-Amz-Algorithm", "AWS4-HMAC-SHA256");
+
+        self.remove_header("X-Amz-Credential");
+        self.remove_header("x-amz-credential");
+        self.params.put("X-Amz-Credential", format!("{}/{}/{}/{}/aws4_request", &creds.aws_access_key_id(), &current_date, self.region.name(), self.service));
+
+        self.remove_header("X-Amz-Expires");
+        self.remove_header("x-amz-expires");
+        let expiration_time = {
```

Comment:
Still calls `.to_string()` unconditionally but rarely uses the resulting string. In my first review, I missed that you got a reference here and `.unwrap_or_else()` won't work but something like this should work:

```rust
let expiration_time = match self.params.get("response-expires") {
            Some(Some(value)) => value.to_string(),
            _ => "3600".to_string(),
};
```